### ISSUE TITLE
Demeteorizing Vulcan and how you can help

### ISSUE THREAD

**Issue by eric-burel** [VulcanJS/vulcan-next-starter#1]
## Goal

Progressively transition from Meteor to Next app.

- First being able to connect any type of JS frontend to a Vulcan backend
- Then reuse Vulcan code in Next, both backend and frontend
- Then use Next as our main framework

Some code will be specific to Next, however this transition will also allow to reuse Vulcan code in other type of React frameworks (eg Gatsby), at least for the frontend.

## Things to get sorted out
- [ ] How to create `npm` packages that works differently on the client and the server? The idea is to provide fullstack package as we are used to do in Vulcan. For Storybook [I've created a specific Webpack handler](https://www.npmjs.com/package/vulcan-loader), but it would be better to have some out-of-the-box pattern.
- [ ] How to handle the Mongo database
- [ ] How to handle startup logic. On the client it's ok (see Storybook code in Vulcan) but on the server we don't really have a startup system. We might probably explicitely call a function in our Next app.
- [ ] ... others

### From Meteor packages to NPM packages
- [ ] Create a Lerna monorepo to store Vulcan packages
- [ ] Copy Vulcan Meteor packages into this repo
- [ ] Demeteorize them so they work with NPM
- [ ] Update the Meteor packages to reuse the NPM packages as much as possible => from `meteor/vulcan-core` to `@vulcan/core`.

The goal is to both allow to use Vulcan outside Meteor, in Next, and keep maintaining the Meteor version with a minimal amount of code.

### Specific demeteorization tasks

- [ ] Move to Mongoose for collections https://github.com/VulcanJS/Vulcan/issues/2479
- [ ] Graphql Accounts [issue](https://github.com/VulcanJS/Vulcan/issues/1896) [PR](https://github.com/VulcanJS/Vulcan/pull/2530)

## Use case idea

Working on a simple business use case would allow us to create example more easily. We could for example develop a Job bord, or an app similar to "Awesome Vulcan" to publish vulcan ressources.

[This document is a Work in Progress. It will be updated with additional questions, answers, and process to move from Meteor to Next]

**Comment by comus** [comment]
hello, something i tried to do it last year, but i didn't have time to finish it. really.

i just share what i did, for the "How to create npm packages that works differently on the client and the server?"

it s easier with webpack config. it's good to wrap it with next.config.js, just as `withVulcan`

`config.resolve.mainFiles`

for example
https://github.com/comus/react-vulcan-proposal/blob/master/src/withVulcan.js

and "Mongo database", i am not use mongodb now. but i have a solution for local development.

there is a node js library called 'mongodb-prebuilt', it can download the mongodb binary and run it.

i have a bin script: https://github.com/comus/react-vulcan-proposal/blob/master/src/bin/react-vulcan.js

with the bin, people can just type `npx @vulcan/anyname mongo` to launch a mongodb for development.

**Comment by eric-burel** [comment]
Thanks for the `withVulcan` helper. So it seems that you went for a specific loader too in order to handle both client and server. Maybe that's the only solution indeed.

**Comment by eric-burel** [comment]
I've made a first update of the monorepo to show how it could work: https://github.com/VulcanJS/vulcan-npm

I've only added Typescript, but we would also need to use Webpack in order to have transparent usage between server and client. It's not tested yet.

@comus I have reused your idea of using `mainFiles` to make a more generic helper in `@vulcan/webpack` package. It may be used in Next like you did but also in any kind of app that uses Webpack.

Since Node servers use Webpack less often, we could also consider that `index.js` is the default server export. So that you can still do  `const whatever = require('@vulcan/core')` in a server that do not use Webpack, it would be equivalent to `const whatever = require('@vulcan/core/index.server')`.
 
We can safely consider that client side the developer is able to setup Webpack to import Vulcan, or explicitely call `const whatever = require('@vulcan/core/index.client`).

Don't know if I am clear but I think we will figure this out correctly.

**Comment by eric-burel** [comment]
@comus I've pushed a new test, sadly I could'nt have "multi-env" import to work. I have pushed a new commit with a demonstration (see `_app.ts`, it loads `@vulcan/multi-env-demo`).
It is weird because the "log" is correct, but I still have a build error, very weird.

Also, I made a test with Webpack in the monorepo and build did work. However we face a big problem: the text editor can't understand our `index.client`/`index.server` thing. It says that it does not find the package if you just type `@vulcan/multi-env-demo`, because it looks for an `index`, even when the build work.
There is no way for VS code to tell whether you are editing client code or server code. Note that it explains poor support of Meteor in text editors.

This may end up being too complicated. In which case we would be forced to do `import '@vulcan/core/server` and `import '@vulcan/core/client` depending on the environment, without relying on Weback.

## To sum it up:

- `@vulcan/core` would import only code that works on BOTH client and server
- `@vulcan/core/server` code that works on BOTH client and server + code that work on server only
- `@vulcan/core/server` code that works on BOTH client and server + code that work on client only

The example with `resolvemainFiles` try to be a bit smarter and select the right code automatically but it may raise too many issues.

**Comment by eric-burel** [comment]
It also means that we will lose the pattern of having 2 components with same name but different implementation in server and client, eg to handle components that do not support SSR.
But again not a big deal, it was seldom used and very Vulcan specific, we have other alternatives in Next (using NoSSR components for example, dynamic imports and so on)
